src/t_hash.c

```diff
@@ -2030,3 +2074,165 @@ void hrandfieldCommand(client *c) {
+}
+
+/* Propagate hash field deletions to replicas and AOF. */
+static void propagateFieldsDeletion(serverDb *db, robj **argv, int argc) {
+    /* If the primary decided to delete a hash field we must propagate it to replicas no matter what.
+     * Even if module executed a command without asking for propagation. */
+    int prev_replication_allowed = server.replication_allowed;
+    server.replication_allowed = 1;
+    alsoPropagate(db->id, argv, argc, PROPAGATE_AOF | PROPAGATE_REPL);
+    robj *keyobj = argv[1];
+    notifyKeyspaceEvent(NOTIFY_EXPIRED, "hexpired", keyobj, db->id);
+
+    server.replication_allowed = prev_replication_allowed;
+}
+
+/* Callback for popping expired entries from the volatile set.
+ * Deletes the entry from the hash table and tracks it in the expiry context.
+ * Returns 1 if deleted, 0 if nothing to do.
+ */
+static int expireEntry(void *entry, void *c) {
+    expiryContext *ctx = c;
+    robj *o = ctx->key;
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+
+    hashtable *ht = o->ptr;
+    void *entry_ptr = NULL;
+    int deleted = hashtablePop(ht, entry, &entry_ptr);
+
+    if (deleted) {
+        debugLogField(o, entry_ptr);
+        ctxAddEntry(ctx, entry_ptr);
+        server.stat_expiredfields++;
+        return 1;
+    }
+    return 0;
+}
+
+/* Free all entry memory for a given list of expired entries. */
+void freeEntries(void **entries, int n) {
+    for (int i = 0; i < n; i++) {
+        entryFree(entries[i]);
+    }
+}
+
+/* Free all robj references created in an argv list, except the shared command in slot 0. */
+void freeArgvObjects(robj **argv, int argc) {
+    for (int i = 1; i < argc; i++) {
+        // skip 1, the shared command
+        decrRefCount(argv[i]);
+    }
+}
+
+/* Build argv array for propagating HDEL of expired hash fields.
+ * argv[0] = shared.hdel
+ * argv[1] = key object
+ * argv[2..] = field names to delete
+ * Allocates a new key object, but *does not* free it — the caller must later.
+ * Returns the final argc.
+ */
+static int buildExpireFieldsArgv(void **entries, int n_entries, robj *o, robj *argv[]) {
+    int argc = 0;
+    robj *keyobj = createStringObjectFromSds(objectGetKey(o));
+    argv[argc++] = shared.hdel; // HDEL command
+    argv[argc++] = keyobj;      // key name
+    for (int i = 0; i < n_entries; i++) {
+        // field to delete
+        argv[argc++] = createStringObjectFromSds(entryGetField(entries[i]));
+    }
+    return argc;
+}
+
+/* Process expired fields for a hash key, deleting them,
+ * and propagating changes to replicas and AOF.
+ *
+ * This routine:
+ *  - identifies expired hash fields from a volatile set
+ *  - deletes them
+ *  - frees the entire key if the hash is empty
+ *  - propagates HDEL commands if the hash object isn't empty, and propagates hash DEL if empty.
+ *
+ * Returns the number of expired fields removed.
+ */
+size_t activeExpireFieldProcessKey(robj *o, serverDb *db, mstime_t now, unsigned long max_entries) {
```

Comment:
I think we should stick to conventions. if this is a hash function it should be called like it. For example:
```suggestion
size_t hashTypeReclaimExpireFields(robj *o, serverDb *db, mstime_t now, unsigned long max_entries) {
```